### PR TITLE
Fixes for Julia 0.6

### PR DESCRIPTION
**Pull request by fredrikekre**
putting this branch here now instead

### CODE REVIEW

**Comment by KristofferC** [comment]
https://github.com/JuliaDiffEq/FiniteElementDiffEq.jl/blob/master/.travis.yml#L10 for allow fails

**Comment by fredrikekre** [comment]
should have rebased on new `.yml` files first :p

**Comment by KristofferC** [comment]
i did it for u brah

**Comment by fredrikekre** [comment]
Moving this comment here:

StaticArrays works now on master.

- [ ] Should we remove [dot-operations](https://github.com/KristofferC/ContMechTensors.jl/blob/master/src/ContMechTensors.jl#L159)? That give deprecation warning now.
- [ ] On `v0.6` [this](https://github.com/KristofferC/ContMechTensors.jl/blob/master/src/ContMechTensors.jl#L205) method is more specific than [this](https://github.com/KristofferC/ContMechTensors.jl/blob/master/src/ContMechTensors.jl#L183), how come? That causes some test to fail too:

```julia
julia> ones(Vec{3})
ERROR: ArgumentError: `one` only defined for order 2 and 4
Stacktrace:
 [1] one(...) at /home/fredrik/Dropbox/Programming/ContMechTensors.jl/src/ContMechTensors.jl:237
 [2] ones(::Type{ContMechTensors.Tensor{1,3,T,3} where T}) at /home/fredrik/Dropbox/Programming/ContMechTensors.jl/src/ContMechTensors.jl:190
```

**Comment by fredrikekre** [comment]
Deprecate `ones` and add a method for `fill` instead? That would also solve point 2 above.

**Comment by fredrikekre** [comment]
`fill` could also include `zero`, but that is a quite useful shortcut of `fill(0.0, Tensor{2, 3})`

**Comment by KristofferC** [comment]
I think we should try to adhere to Base arrays, and if you can do `ones` on them it would be good if it could be done here too. We could add `fill` and just make `ones` and `zeros` call `fill`?

**Comment by fredrikekre** [comment]
Sure. What's up with point 2 though?

**Comment by fredrikekre** [comment]
Ah... `Vec` includes the `M` argument in the type. But why do this work on v0.5?

**Comment by fredrikekre** [comment]
Edit: Today it works... Wut.

**Comment by KristofferC** [comment]
0.6, faster than the wind

**Comment by KristofferC** [comment]
This is not testing on nightly

**Comment by fredrikekre** [comment]
I mean locally

**Comment by fredrikekre** [comment]
Edit2: It doesn't work. Calling `ones(Vec{3, Float64})` is equal to `ones(Tensor{1, 3, Float64, 3})`. Why does that dispatch to [this](https://github.com/KristofferC/Tensors.jl/blob/master/src/Tensors.jl#L205) instead of [this](https://github.com/KristofferC/Tensors.jl/blob/master/src/Tensors.jl#L182)? I don't understand....

**Comment by KristofferC** [comment]
Not sure.. we should try make a small example and report it.

File: src/Tensors.jl [hunk 1]
```diff
@@ -174,6 +172,16 @@ end
 # Basic constructors #
 ######################
 
+# fill
```
Comment (KristofferC) [inline src/Tensors.jl]: U wot m8?

**Comment by fredrikekre** [comment]
```julia
julia> A = rand(Tensor{2, 2})
2×2 Tensors.Tensor{2,2,Float64,4}:
Error showing value of type Tensors.Tensor{2,2,Float64,4}:
ERROR: indexing not defined for Tensors.Tensor{2,2,Float64,4}
Stacktrace:
 [1] getindex(::Tensors.Tensor{2,2,Float64,4}, ::Int64, ::Int64) at ./abstractarray.jl:863
 [2] isassigned(::Tensors.Tensor{2,2,Float64,4}, ::Int64, ::Int64, ::Vararg{Int64,N} where N) at ./abstractarray.jl:214
 [3] alignment(::IOContext{Base.Terminals.TTYTerminal}, ::Tensors.Tensor{2,2,Float64,4}, ::Array{Int64,1}, ::Array{Int64,1}, ::Int64, ::Int64, ::Int64) at ./show.jl:1344
 [4] print_matrix(::IOContext{Base.Terminals.TTYTerminal}, ::Tensors.Tensor{2,2,Float64,4}, ::String, ::String, ::String, ::String, ::String, ::String, ::Int64, ::Int64) at ./show.jl:1474
 [5] print_matrix(::IOContext{Base.Terminals.TTYTerminal}, ::Tensors.Tensor{2,2,Float64,4}, ::String, ::String, ::String) at ./show.jl:1446
 [6] #showarray#254(::Bool, ::Function, ::IOContext{Base.Terminals.TTYTerminal}, ::Tensors.Tensor{2,2,Float64,4}, ::Bool) at ./show.jl:1696
 [7] display(::Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, ::MIME{Symbol("text/plain")}, ::Tensors.Tensor{2,2,Float64,4}) at ./REPL.jl:122
 [8] display(::Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, ::Tensors.Tensor{2,2,Float64,4}) at ./REPL.jl:125
 [9] display(::Tensors.Tensor{2,2,Float64,4}) at ./multimedia.jl:194
```

File: src/Tensors.jl (hunk 1 above)
Comment (fredrikekre) [inline src/Tensors.jl]: woops

File: src/Tensors.jl [hunk 2]
```diff
@@ -156,8 +154,8 @@ for TensorType in (SymmetricTensor, Tensor)
     end
 end
 
-# Binary op(Tensor, Tensor): +, -, .+, .-, .*, ./
-for op in (:+, :-, :.+, :.-, :.*, :./)
+# Binary op(Tensor, Tensor): +, -
+for op in (:+, :-)
```
Comment (fredrikekre) [inline src/Tensors.jl]: This shouldnt be in this pr either.

Comment (fredrikekre) [inline src/Tensors.jl]: Oh wait, it should be. Can we drop dot operators?

**Comment by fredrikekre** [comment]
The comment above fails all tests since it cannot convert to Array. What is wrong with the getindex implementation?

File: src/Tensors.jl (hunk 2 above)
Comment (KristofferC) [inline src/Tensors.jl]: Probably, not sure if someone wants to do `.^` for example.

Comment (fredrikekre) [inline src/Tensors.jl]: That was not supported anyway.
No matter what, these should be implemented with `broadcast`. Can be a separate PR.

**Comment by KristofferC** [comment]
```
julia> Base.linearindexing(Tensor)
Base.LinearSlow()
```

**Comment by KristofferC** [comment]
Need to update my julia..

**Comment by KristofferC** [comment]
@ompat`?

**Comment by fredrikekre** [comment]
I get this on 0.5:
```julia
julia> using Tensors
LoadError: ------ UndefVarError ------------------- Stacktrace (most recent call last)

 [1] — include_from_node1(::String) at loading.jl:488

UndefVarError: IndexStyle not defined
while loading /home/fredrik/Dropbox/Programming/Tensors.jl/src/Tensors.jl, in expression starting on line 90
```

**Comment by fredrikekre** [comment]
on master:

```julia
julia> include("test/runtests.jl")
Test Summary:                        | Pass  Total
basic constructors: rand, zero, ones |  654    654
Test Summary: | Pass  Total
diagm, one    |  532    532
Test Summary: | Pass  Total
base vectors  |   30     30
Test Summary: | Pass  Total
simple math   |  390    390
Test Summary:          | Pass  Total
create with a function |  396    396
create from Array: Error During Test
  Got an exception of type MethodError outside of a @test
  MethodError: no method matching n_components(::Type{Tensors.Tensor{2,1,T,M} where M where T<:Real})
  The applicable method may be too new: running in world age 21579, while current world is 21821.
  Closest candidates are:
    n_components(::Type{Tensors.Tensor{order,dim,T,M} where M where T<:Real}) where {order, dim} at /home/fredrik/Dropbox/Programming/Tensors.jl/src/Tensors.jl:78 (method too new to be called from this world context.)
    n_components(::Type{Tensors.SymmetricTensor{2,dim,T,M} where M where T<:Real}) where dim at /home/fredrik/Dropbox/Programming/Tensors.jl/src/Tensors.jl:73 (method too new to be called from this world context.)
    n_components(::Type{Tensors.SymmetricTensor{4,dim,T,M} where M where T<:Real}) where dim at /home/fredrik/Dropbox/Programming/Tensors.jl/src/Tensors.jl:75 (method too new to be called from this world context.)
  Stacktrace:
   [1] compute_index(::Type{Tensors.Tensor{4,1,T,M} where M where T<:Real}, ::Int64, ::Int64, ::Int64, ::Int64) at /home/fredrik/Dropbox/Programming/Tensors.jl/src/indexing.jl:22
   [2] minorsymmetric(...) at /home/fredrik/Dropbox/Programming/Tensors.jl/src/symmetric.jl:55
   [3] symmetric(::Tensors.Tensor{4,1,Float32,1}) at /home/fredrik/Dropbox/Programming/Tensors.jl/src/symmetric.jl:74
   [4] macro expansion at /home/fredrik/Dropbox/Programming/Tensors.jl/test/test_misc.jl:221 [inlined]
   [5] macro expansion at ./test.jl:853 [inlined]
   [6] anonymous at ./<missing>:?
   [7] include_from_node1(::String) at ./loading.jl:539
   [8] include(::String) at ./sysimg.jl:14
   [9] include_from_node1(::String) at ./loading.jl:539
   [10] include(::String) at ./sysimg.jl:14
   [11] eval(::Module, ::Any) at ./boot.jl:235
   [12] eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:66
   [13] macro expansion at ./REPL.jl:97 [inlined]
   [14] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
Test Summary:     | Pass  Error  Total
create from Array |   52      1     53
ERROR: LoadError: LoadError: Some tests did not pass: 52 passed, 0 failed, 1 errored, 0 broken.
while loading /home/fredrik/Dropbox/Programming/Tensors.jl/test/test_misc.jl, in expression starting on line 207
while loading /home/fredrik/Dropbox/Programming/Tensors.jl/test/runtests.jl, in expression starting on line 4
```

**Comment by KristofferC** [comment]
Too old compat on v0.5?

**Comment by KristofferC** [comment]
Because of OhMyREPL lol!

**Comment by fredrikekre** [comment]
No

**Comment by fredrikekre** [comment]
Oh...

**Comment by KristofferC** [comment]
I get:

```jl
julia> include("/home/kristoffer/Dropbox/julia/Tensors/test/runtests.jl")
Test Summary:                        | Pass  Total
basic constructors: rand, zero, ones |  654    654
Test Summary: | Pass  Total
diagm, one    |  532    532
Test Summary: | Pass  Total
base vectors  |   30     30
Test Summary: | Pass  Total
simple math   |  390    390
Test Summary:          | Pass  Total
create with a function |  396    396
Test Summary:     | Pass  Total
create from Array |  168    168
Test Summary: | Pass  Total
indexing      | 1806   1806
Test Summary:              | Pass  Total
norm, trace, det, inv, eig |  174    174
Test Summary:     | Pass  Total
tensor identities |  192    192
Test Summary:        | Pass  Total
promotion/conversion |  256    256
Test Summary: | Pass  Total
exceptions    |    8      8
Test Summary:     | Pass  Total
tensor operations |  884    884
WARNING: Base.LinearFast is deprecated, use Base.IndexLinear instead.
  likely near /home/kristoffer/Dropbox/julia/Tensors/test/test_ad.jl:23
WARNING: Base.LinearFast is deprecated, use Base.IndexLinear instead.
  likely near /home/kristoffer/Dropbox/julia/Tensors/test/test_ad.jl:23
WARNING: Base.LinearFast is deprecated, use Base.IndexLinear instead.
  likely near /home/kristoffer/Dropbox/julia/Tensors/test/test_ad.jl:23
Testing AD for dim = 1
AD: Error During Test
  Test threw an exception of type MethodError
  Expression: $(Expr(:escape, :(∇(I2, A)))) $(Expr(:escape, :≈)) $(Expr(:escape, :((∇(I2, A, :all))[1]))) $(Expr(:escape, :≈)) $(Expr(:escape, :(DI2(A))))
```

on latest staticarrays.

**Comment by fredrikekre** [comment]
Please abort some builds in the queue

**Comment by fredrikekre** [comment]
This passes for me on v0.5 now.

**Comment by fredrikekre** [comment]
The worldage problem above is failing this on master now.

**Comment by KristofferC** [comment]
Might be a problem in static arrays..

File: .travis.yml [hunk 3]
```diff
@@ -6,7 +6,9 @@ os:
 
 julia:
   - 0.5
-  # - nightly
+matrix:
+  allow_failures:
+    -julia: nightly
```
Comment (fredrikekre) [inline .travis.yml]: nightly kör ju inte ens på travis.

**Comment by KristofferC** [comment]
https://github.com/KristofferC/Tensors.jl/pull/2#issuecomment-279642003

**Comment by KristofferC** [comment]
Seems to work. Sweet!

**Comment by KristofferC** [comment]
Except a few doctests. I guess we could only run them on nightly?

**Comment by fredrikekre** [comment]
Doctest fail is the space between element `(1,2)` -> `(1, 2)`. Will update.

**Comment by fredrikekre** [comment]
Hmmm.. On master I get this. Something changed from the nightly since now?

```julia
julia> include("test/runtests.jl")
basic constructors: rand, zero, ones: Error During Test
  Got an exception of type MethodError outside of a @test
  MethodError: no method matching get_type(::Type{Type{Tensors.Tensor{1,1,T,M} where M where T<:Real}})
  The applicable method may be too new: running in world age 21610, while current world is 21858.
  Closest candidates are:
    get_type(::Type{Type{X}}) where X at /home/fredrik/Dropbox/Programming/Tensors.jl/src/Tensors.jl:70 (method too new to be called from this world context.)
  Stacktrace:
   [1] fill(...) at /home/fredrik/Dropbox/Programming/Tensors.jl/src/constructors.jl:42
   [2] rand(::Type{Tensors.Tensor{1,1,T,M} where M where T<:Real}) at /home/fredrik/Dropbox/Programming/Tensors.jl/src/constructors.jl:32
   [3] macro expansion at /home/fredrik/Dropbox/Programming/Tensors.jl/test/test_misc.jl:7 [inlined]
   [4] macro expansion at ./test.jl:853 [inlined]
   [5] anonymous at ./<missing>:?
   [6] include_from_node1(::String) at ./loading.jl:539
   [7] include(::String) at ./sysimg.jl:14
   [8] include_from_node1(::String) at ./loading.jl:539
   [9] include(::String) at ./sysimg.jl:14
   [10] eval(::Module, ::Any) at ./boot.jl:235
   [11] eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:66
   [12] macro expansion at ./REPL.jl:97 [inlined]
   [13] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
Test Summary:                        | Error  Total
basic constructors: rand, zero, ones |     1      1
ERROR: LoadError: LoadError: Some tests did not pass: 0 passed, 0 failed, 1 errored, 0 broken.
while loading /home/fredrik/Dropbox/Programming/Tensors.jl/test/test_misc.jl, in expression starting on line 1
while loading /home/fredrik/Dropbox/Programming/Tensors.jl/test/runtests.jl, in expression starting on line 4
```

**Comment by fredrikekre** [comment]
Compiled the same commit as nightly and the tests fail locally...

File: test/runtests.jl [hunk 4]
```diff
@@ -5,5 +5,7 @@ include("test_misc.jl")
 include("test_ops.jl")
 include("test_ad.jl")
 
-# Build the docs
```
Comment (KristofferC) [inline test/runtests.jl]: Don't call `exit()` just do.

```
if get(ENV, "TRAVIS_OS_NAME", "")       == "linux" &&
        get(ENV, "TRAVIS_JULIA_VERSION", "") == "0.6"
   include(...)
```

Comment (KristofferC) [inline test/runtests.jl]: Hmm lets do like this instead:

in `make.jl` change `strict = true` to `strict = VERSION.minor == 6`  or something.

Comment (fredrikekre) [inline test/runtests.jl]: Yea, but then the tests will not run if you test locally...

Comment (fredrikekre) [inline test/runtests.jl]: I think its better to always run them, but only in strict mode on that build bot.

Comment (fredrikekre) [inline test/runtests.jl]: Then we should deploy the docs from the 0.6 bot too.

**Comment by fredrikekre** [comment]
Works like this:

```
fredrik@fredrik-laptop:~/Dropbox/Programming/Tensors.jl$ julia-master /home/fredrik/.julia/v0.6/Tensors/test/runtests.jl 
Test Summary:                        | Pass  Total
basic constructors: rand, zero, ones |  654    654
Test Summary: | Pass  Total
diagm, one    |  532    532
Test Summary: | Pass  Total
base vectors  |   30     30
```

**Comment by KristofferC** [comment]
Tag a new version?
